Documentation/design-docs/multi-reg-call-nodes.md

```diff
@@ -0,0 +1,211 @@
+   b) Create 2-4 temp locals matching the type of respective return registers of GT_CALL and 
+      use them to create following IR
+
+      GT_STORE_MULTI_VAR(TYP_STRUCT, <tmpLcl1, tmpLcl2, tmpLcl3, tmpLcl4>, op1 = GT_CALL)
+      IR nodes to extract/assemble fields of tmp struct from individual tmpLcl[1..4]
+
+Platform agnostic ReturnTypeDesc on a GT_CALL node
+---------------------------------------------------
+Every GT_CALL node will have a light-weight and platform independent ReturnTypeDesc that specifies
+ - Respective return types of a multi-reg return value
+ - Respective Return register numbers in which the value is returned.
+ - Actual size of the return value.
+ - A spill mask for LSRA/codegen purpose indicating which of the retun registers needs to be spilled/reloaded.
+   Spil mask provides flexibility to spill/reload selective return registers, without which we would have to
+   spill/reload all return regs together.
+
+ReturnTypeDesc is initialized during importation while creating GenTreeCall node.
+
+
+Post Global Morph, IR forms of GT_RETURN(tmp) where tmp is of TYP_STRUCT
+-----------------------------------------------------------------------
+Case 3 is morphed into
+
+    GT_RETURN (TYP_STRUCT, op1 = GT_MULTI_VAR(TYP_STRUCT, <Fieldlcl1, FieldLcl2, FieldLcl3, FieldLcl4>))
+    
+    Where FieldLcl[1..4] are lcl numbers of independently promoted fields of tmp and
+    GT_MULTI_VAR is a new node that represents 2-4 independent local vars.
+
+Case 1 remains unchanged
+    
+    GT_RETURN(TYP_STRUCT, op1 = tmp)    
+
+Case 2 is handled as follows:
+    a) Force tmp to memory and morph it as in case 1 above
+    b) Create 2-4 temp locals matching the type of respective return registers of GT_RETURN and
+       use them to extract individual fields from tmp and morph it as in case 3 above.
+
+          tmpLcl1 = GenTree Nodes to extract first 8-bytes from tmp 
+          tmpLcl2 = GenTree Nodes to extract next 8-bytes from tmp and so on
+           
+          GT_RETURN(TYP_STRUCT, op1 = GT_STORE_MULTI_VAR(TYP_STRUCT, <tmpLcl1, tmpLcl2, tmpLcl3, tmpLcl4>, tmp))
+
+
+Post Lowering, IR forms fo tmp=GT_CALL where tmp is of TYP_LONG
+--------------------------------------------------------------
+TYP_LONG lcl vars are promoted and decomposed in IR lowering phase. The equivalent of global morphing done
+for struct types need to be done for long types during IR Lowering phase.  
+
+For e.g. tmp = GT_CALL post rationlization will be GT_STORE_LCL_VAR(TYP_LONG, lcl Num of tmp, op1 = GT_CALL).
+Post IR lowering, it will be transformed into
+
+     GT_STORE_MULTI_VAR(TYP_LONG, <tmpHi, tmpLo>, op1=GT_CALL)  
+     GT_LONG(GT_LCL_VAR(tmpHi), GT_LCL_VAR(tmpLo))
+
+     where tmpHi and tmpLo are promoted lcl fields of tmp.  
+
+If tmp cannot be promoted, we will create two new tmps of TYP_INT and transform it into
+
+    GT_STORE_MULTI_VAR(TYP_LONG, <tmpLcl1, tmpLcl2>, op1=GT_CALL)
+    GT_LONG(GT_LCL_VAR(tmpLcl1), GT_LCL_VAR(tmpLcl2))
+      
```

Comment:
This is the case where tmp is not itself a "candidate" (i.e. it is marked doNotEnregister, or it is not tracked). That should be clarified - lower doesn't really make any "promotion" decisions per se, it simply "promotes" the variable if and only if it is a candidate.
If it is not, then I think that we should handle it as for the structs Case 1: It would be GT_STORE(tmp), and the store would consume the two registers.